File: Samples/Minjie_ARCHIVE_2_0/Minjie/Minjie/ScreenManager/MenuEntry.cs
#region File Description
//-----------------------------------------------------------------------------
// MenuEntry.cs
//
// XNA Community Game Platform
// Copyright (C) Microsoft Corporation. All rights reserved.
//-----------------------------------------------------------------------------
#endregion

#region Using Statements
using System;
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
#endregion

namespace Minjie
{
    /// <summary>
    /// Helper class represents a single entry in a MenuScreen. This also provides 
    /// an event that will be raised when the menu entry is selected.
    /// </summary>
    /// <remarks>Based on a similar class in the Game State Management sample.</remarks>
    class MenuEntry
    {
        #region Fields

        /// <summary>
        /// The texture rendered for this entry.
        /// </summary>
        Texture2D texture;

        /// <summary>
        /// The origin of the texture.
        /// </summary>
        Vector2 origin;

        /// <summary>
        /// Tracks a fading selection effect on the entry.
        /// </summary>
        /// <remarks>
        /// The entries transition out of the selection effect when they are deselected.
        /// </remarks>
        float selectionFade;

        #endregion

        #region Properties


        /// <summary>
        /// Gets or sets the texture of this menu entry.
        /// </summary>
        public Texture2D Texture
        {
            get { return texture; }
            set 
            {
                texture = value;
                if (texture == null)
                {
                    origin = Vector2.Zero;
                }
                else
                {
                    origin = new Vector2(texture.Width / 2, texture.Height / 2);
                }
            }
        }


        #endregion

        #region Events


        /// <summary>
        /// Event raised when the menu entry is selected.
        /// </summary>
        public event EventHandler<EventArgs> Selected;


        /// <summary>
        /// Method for raising the Selected event.
        /// </summary>
        protected internal virtual void OnSelectEntry()
        {
            if (Selected != null)
                Selected(this, EventArgs.Empty);
        }


        #endregion

        #region Initialization


        /// <summary>
        /// Constructs a new menu entry.
        /// </summary>
        public MenuEntry() { }


        #endregion

        #region Update and Draw


        /// <summary>
        /// Updates the menu entry.
        /// </summary>
        public virtual void Update(MenuScreen screen, bool isSelected,
                                                      GameTime gameTime)
        {
            // When the menu selection changes, entries gradually fade between
            // their selected and deselected appearance, rather than instantly
            // popping to the new state.
            float fadeSpeed = (float)gameTime.ElapsedGameTime.TotalSeconds * 4;

            if (isSelected)
                selectionFade = Math.Min(selectionFade + fadeSpeed, 1);
            else
                selectionFade = Math.Max(selectionFade - fadeSpeed, 0);
        }


        /// <summary>
        /// Draws the menu entry. This can be overridden to customize the appearance.
        /// </summary>
        public virtual void Draw(MenuScreen screen, Vector2 position,
                                 bool isSelected, GameTime gameTime)
        {
            // Pulsate the size of the selected menu entry.
            double time = gameTime.TotalGameTime.TotalSeconds;
            
            float pulsate = (float)Math.Sin(time * 6) + 1;
            
            float scale = 0.7f + pulsate * 0.05f * selectionFade;

            // Draw texture, centered on the screen
            ScreenManager screenManager = screen.ScreenManager;
            SpriteBatch spriteBatch = screenManager.SpriteBatch;

            if (texture != null)
            {
                spriteBatch.Draw(texture, position, null, Color.White, 0f, origin, 
                    scale, SpriteEffects.None, 0);
            }
        }


        /// <summary>
        /// Queries how much space this menu entry requires.
        /// </summary>
        public virtual int GetHeight(MenuScreen screen)
        {
            return (int)Math.Floor(texture.Height * 0.8f);
        }


        #endregion
    }
}

File: Samples/GSMSample_4_0_Mango_VB/GameStateManagementSample/GameStateManagementSample/Screens/MenuEntry.vb
#Region "File Description"
'-----------------------------------------------------------------------------
' MenuEntry.vb
'
' XNA Community Game Platform
' Copyright (C) Microsoft Corporation. All rights reserved.
'-----------------------------------------------------------------------------
#End Region

#Region "Using Statements"
Imports Microsoft.VisualBasic
Imports Microsoft.Xna.Framework
Imports Microsoft.Xna.Framework.Graphics
Imports GameStateManagement
#End Region

''' <summary>
''' Helper class represents a single entry in a MenuScreen. By default this
''' just draws the entry text string, but it can be customized to display menu
''' entries in different ways. This also provides an event that will be raised
''' when the menu entry is selected.
''' </summary>
Friend Class MenuEntry
#Region "Fields"

    ''' <summary>
    ''' The text rendered for this entry.
    ''' </summary>
    Private _text As String

    ''' <summary>
    ''' Tracks a fading selection effect on the entry.
    ''' </summary>
    ''' <remarks>
    ''' The entries transition out of the selection effect when they are deselected.
    ''' </remarks>
    Private selectionFade As Single

    ''' <summary>
    ''' The position at which the entry is drawn. This is set by the MenuScreen
    ''' each frame in Update.
    ''' </summary>
    Private position As Vector2

#End Region

#Region "Properties"


    ''' <summary>
    ''' Gets or sets the text of this menu entry.
    ''' </summary>
    Public Property Text() As String
        Get
            Return _text
        End Get
        Set(ByVal value As String)
            _text = value
        End Set
    End Property


    ''' <summary>
    ''' Gets or sets the position at which to draw this menu entry.
    ''' </summary>
    Public Property _Position() As Vector2

#End Region

#Region "Events"


    ''' <summary>
    ''' Event raised when the menu entry is selected.
    ''' </summary>
    Public Event Selected As EventHandler(Of PlayerIndexEventArgs)


    ''' <summary>
    ''' Method for raising the Selected event.
    ''' </summary>
    Protected Friend Overridable Sub OnSelectEntry(ByVal playerIndex As PlayerIndex)
        RaiseEvent Selected(Me, New PlayerIndexEventArgs(playerIndex))
    End Sub


#End Region

#Region "Initialization"


    ''' <summary>
    ''' Constructs a new menu entry with the specified text.
    ''' </summary>
    Public Sub New(ByVal text As String)
        Me._text = text
    End Sub


#End Region

#Region "Update and Draw"


    ''' <summary>
    ''' Updates the menu entry.
    ''' </summary>
    Public Overridable Sub Update(ByVal screen As MenuScreen, ByVal isSelected As Boolean, ByVal gameTime As GameTime)
        ' there is no such thing as a selected item on Windows Phone, so we always
        ' force isSelected to be false
#If WINDOWS_PHONE Then
			isSelected = False
#End If

        ' When the menu selection changes, entries gradually fade between
        ' their selected and deselected appearance, rather than instantly
        ' popping to the new state.
        Dim fadeSpeed As Single = CSng(gameTime.ElapsedGameTime.TotalSeconds) * 4

        If isSelected Then
            selectionFade = Math.Min(selectionFade + fadeSpeed, 1)
        Else
            selectionFade = Math.Max(selectionFade - fadeSpeed, 0)
        End If
    End Sub


    ''' <summary>
    ''' Draws the menu entry. This can be overridden to customize the appearance.
    ''' </summary>
    Public Overridable Sub Draw(ByVal screen As MenuScreen, ByVal isSelected As Boolean, ByVal gameTime As GameTime)
        ' there is no such thing as a selected item on Windows Phone, so we always
        ' force isSelected to be false
#If WINDOWS_PHONE Then
			isSelected = False
#End If

        ' Draw the selected entry in yellow, otherwise white.
        Dim color As Color = If(isSelected, color.Yellow, color.White)

        ' Pulsate the size of the selected menu entry.
        Dim time As Double = gameTime.TotalGameTime.TotalSeconds

        Dim pulsate As Single = CSng(Math.Sin(time * 6)) + 1

        Dim scale As Single = 1 + pulsate * 0.05F * selectionFade

        ' Modify the alpha to fade text out during transitions.
        color *= screen.TransitionAlpha

        ' Draw text, centered on the middle of each line.
        Dim screenManager As ScreenManager = screen.ScreenManager
        Dim spriteBatch As SpriteBatch = screenManager.SpriteBatch
        Dim font As SpriteFont = screenManager.Font

        Dim origin As New Vector2(0, CSng(font.LineSpacing) / 2)

        spriteBatch.DrawString(font, _text, _Position, color, 0, origin, scale, SpriteEffects.None, 0)
    End Sub


    ''' <summary>
    ''' Queries how much space this menu entry requires.
    ''' </summary>
    Public Overridable Function GetHeight(ByVal screen As MenuScreen) As Integer
        Return screen.ScreenManager.Font.LineSpacing
    End Function


    ''' <summary>
    ''' Queries how wide the entry is, used for centering on the screen.
    ''' </summary>
    Public Overridable Function GetWidth(ByVal screen As MenuScreen) As Integer
        Return CInt(screen.ScreenManager.Font.MeasureString(Text).X)
    End Function


#End Region
End Class